File: api/ask.js
import fs from "fs";
import path from "path";

const ROOT_DIR = process.cwd();
const VECTOR_PATH = path.join(ROOT_DIR, "db", "kb_vectors.json");

const EMBEDDING_MODEL =
  process.env.OPENAI_EMBEDDING_MODEL || "text-embedding-3-small";
const GEMINI_MODEL = process.env.GEMINI_MODEL || "gemini-3-flash-preview";

const OPENAI_TIMEOUT = Number(process.env.OPENAI_TIMEOUT_SEC || 6);
const GEMINI_TIMEOUT = Number(process.env.GEMINI_TIMEOUT_SEC || 8);
const MAX_CONTEXT_CHARS = Number(process.env.MAX_CONTEXT_CHARS || 6000);

const SUMMARY_TRIGGERS = [
  "summarize",
  "summary",
  "overview",
  "who are you",
  "about you",
  "tell me about yourself",
  "yourself",
  "background",
];

const TOPIC_TRIGGERS = [
  ["projects", ["project", "projects", "portfolio"]],
  ["experience", ["experience", "work", "job", "roles"]],
  ["skills", ["skills", "stack", "tools", "tech"]],
  ["about", ["education", "degree", "university", "school"]],
  ["research", ["research", "paper", "publication"]],
];

const SENSITIVE_TRIGGERS = ["visa", "sponsorship", "citizenship", "immigration"];

let VECTOR_CACHE = null;

function normalizeWhitespace(text) {
  return text.replace(/\s+/g, " ").trim();
}

function stripFrontmatter(text) {
  const clean = text.replace(/^\uFEFF/, "");
  if (!clean.startsWith("---")) return clean;
  const lines = clean.split("\n");
  let endIdx = -1;
  for (let i = 1; i < lines.length; i += 1) {
    if (lines[i].trim() === "---") {
      endIdx = i;
      break;
    }
  }
  if (endIdx === -1) return clean;
  return lines.slice(endIdx + 1).join("\n").trim();
}

function tokenize(text) {
  return normalizeWhitespace(text.toLowerCase())
    .replace(/[^a-z0-9\s]/g, " ")
    .split(" ")
    .filter(Boolean);
}

function hasEvidence(query, contexts) {
  const queryTokens = new Set(tokenize(query));
  if (!queryTokens.size) return true;
  const joined = contexts.map((c) => c.content).join(" ");
  const contextTokens = new Set(tokenize(joined));
  for (const token of queryTokens) {
    if (contextTokens.has(token)) return true;
  }
  return false;
}

function inferTopic(query) {
  const q = query.toLowerCase();
  for (const [topic, triggers] of TOPIC_TRIGGERS) {
    if (triggers.some((t) => q.includes(t))) return topic;
  }
  return null;
}

function isSummaryIntent(query) {
  const q = query.toLowerCase();
  return SUMMARY_TRIGGERS.some((t) => q.includes(t));
}

function isSensitiveQuery(query) {
  const q = query.toLowerCase();
  return SENSITIVE_TRIGGERS.some((t) => q.includes(t));
}

function cosineSimilarity(vecA, normA, vecB, normB) {
  if (!normA || !normB) return 0;
  let dot = 0;
  const len = Math.min(vecA.length, vecB.length);
  for (let i = 0; i < len; i += 1) {
    dot += vecA[i] * vecB[i];
  }
  return dot / (normA * normB);
}

function loadVectors() {
  if (VECTOR_CACHE) return VECTOR_CACHE;
  if (!fs.existsSync(VECTOR_PATH)) {
    throw new Error(`Vector file not found: ${VECTOR_PATH}`);
  }
  const raw = fs.readFileSync(VECTOR_PATH, "utf-8");
  const data = JSON.parse(raw);
  const vectors = (data.vectors || []).map((item) => {
    const embedding = item.embedding || [];
    let norm = 0;
    for (const val of embedding) norm += val * val;
    norm = Math.sqrt(norm);
    return {
      content: stripFrontmatter(item.content || ""),
      embedding,
      norm,
      metadata: item.metadata || {},
    };
  });
  VECTOR_CACHE = vectors;
  return vectors;
}

async function openaiEmbed(query) {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) throw new Error("OPENAI_API_KEY is not set.");

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), OPENAI_TIMEOUT * 1000);
  try {
    const resp = await fetch("https://api.openai.com/v1/embeddings", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ model: EMBEDDING_MODEL, input: query }),
      signal: controller.signal,
    });
    if (!resp.ok) {
      throw new Error(`OpenAI embeddings failed (${resp.status})`);
    }
    const payload = await resp.json();
    const data = payload.data || [];
    if (!data.length) throw new Error("No embedding returned.");
    return data[0].embedding;
  } finally {
    clearTimeout(timeout);
  }
}

function buildPrompt(query, contexts) {
  const parts = contexts.map((c) => normalizeWhitespace(c.content));
  let contextBlock = parts.join("\n");
  if (contextBlock.length > MAX_CONTEXT_CHARS) {
    contextBlock = contextBlock.slice(0, MAX_CONTEXT_CHARS);
  }
  return [
    `Question: ${query}`,
    "",
    "Context:",
    contextBlock,
    "",
    "Instructions:",
    "- Answer in first person as if you are the portfolio owner.",
    "- Be conversational and concise (2-4 sentences).",
    "- Do not use markdown, bullet lists, or headings.",
    "- If the answer is not in the context, say you don't have that information.",
    "- Do not mention sources, file names, or the word 'context'.",
  ].join("\n");
}

async function retrieveContext(query) {
  const vectors = loadVectors();
  const queryEmbedding = await openaiEmbed(query);
  let queryNorm = 0;
  for (const val of queryEmbedding) queryNorm += val * val;
  queryNorm = Math.sqrt(queryNorm);

  const topic = inferTopic(query);
  const summaryOnly = isSummaryIntent(query);

  const scored = [];
  const seenHashes = new Set();
  for (const item of vectors) {
    const meta = item.metadata || {};
    if (topic && meta.topic !== topic) continue;
    if (summaryOnly && meta.layer !== "summary") continue;

    const hash = meta.content_hash;
    if (hash && seenHashes.has(hash)) continue;
    if (hash) seenHashes.add(hash);

    const sim = cosineSimilarity(
      queryEmbedding,
      queryNorm,
      item.embedding,
      item.norm
    );
    const distance = 1 - sim;
    const layerRank = Number(meta.layer_rank || 3);
    const score = distance + 0.15 * (layerRank / 10);
    scored.push({
      score,
      distance,
      metadata: meta,
      content: item.content,
    });
  }
  scored.sort((a, b) => a.score - b.score);
  return scored.slice(0, 5);
}

async function callGemini(prompt) {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) throw new Error("GEMINI_API_KEY is not set.");
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), GEMINI_TIMEOUT * 1000);
  try {
    const resp = await fetch(
      `https://generativelanguage.googleapis.com/v1beta/models/${GEMINI_MODEL}:generateContent?key=${apiKey}`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          system_instruction: {
            parts: [
              {
                text: [
                  "You are the portfolio owner speaking in first person.",
                  "Be warm, recruiter-friendly, and concise.",
                  "No markdown, no bullet points, no headings.",
                  "If the answer isn't in the provided context, say you don't have that information.",
                  "Paraphrase the context; do not copy sentences verbatim.",
                ].join(" "),
              },
            ],
          },
          contents: [{ role: "user", parts: [{ text: prompt }] }],
          generationConfig: { temperature: 0.4, maxOutputTokens: 512 },
        }),
        signal: controller.signal,
      }
    );
    if (!resp.ok) {
      throw new Error(`Gemini failed (${resp.status})`);
    }
    const data = await resp.json();
    const candidates = data.candidates || [];
    const parts = candidates[0]?.content?.parts || [];
    return parts.map((p) => p.text || "").join("").trim();
  } finally {
    clearTimeout(timeout);
  }
}

async function streamGemini(prompt, res) {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) throw new Error("GEMINI_API_KEY is not set.");

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), GEMINI_TIMEOUT * 1000);
  const resp = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/${GEMINI_MODEL}:streamGenerateContent?key=${apiKey}`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        system_instruction: {
          parts: [
            {
              text: [
                "You are the portfolio owner speaking in first person.",
                "Be warm, recruiter-friendly, and concise.",
                "No markdown, no bullet points, no headings.",
                "If the answer isn't in the provided context, say you don't have that information.",
                "Paraphrase the context; do not copy sentences verbatim.",
              ].join(" "),
            },
          ],
        },
        contents: [{ role: "user", parts: [{ text: prompt }] }],
        generationConfig: { temperature: 0.4, maxOutputTokens: 512 },
      }),
      signal: controller.signal,
    }
  );

  try {
    if (!resp.ok || !resp.body) {
      throw new Error(`Gemini stream failed (${resp.status})`);
    }

    const reader = resp.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";
    let wrote = false;

    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";

      for (let line of lines) {
        line = line.trim();
        if (!line) continue;
        if (line.startsWith("data:")) line = line.slice(5).trim();
        if (line === "[DONE]") return true;

        try {
          const payload = JSON.parse(line);
          const candidates = payload.candidates || [];
          for (const candidate of candidates) {
            const parts = candidate.content?.parts || [];
            for (const part of parts) {
              if (part.text) {
                res.write(part.text);
                res.flush?.();
                wrote = true;
              }
            }
          }
        } catch {
          // ignore parse errors
        }
      }
    }
    return wrote;
  } finally {
    clearTimeout(timeout);
  }
}

export default async function handler(req, res) {
  if (req.method === "OPTIONS") {
    res.status(204);
    res.setHeader("Access-Control-Allow-Origin", "*");
    res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
    res.setHeader("Access-Control-Allow-Headers", "Content-Type");
    res.end();
    return;
  }

  if (req.method !== "POST") {
    res.status(405).json({ error: "Method not allowed" });
    return;
  }

  try {
    const { query } = req.body || {};
    const question = (query || "").trim();
    if (!question) {
      res.status(400).json({ error: "Missing query" });
      return;
    }

    res.setHeader("Content-Type", "text/plain; charset=utf-8");
    res.setHeader("Cache-Control", "no-cache, no-transform");
    res.setHeader("X-Gemini-Model", GEMINI_MODEL);

    const contexts = await retrieveContext(question);
    const summaryIntent = isSummaryIntent(question);
    const topic = inferTopic(question);

    if (isSensitiveQuery(question) && !hasEvidence(question, contexts)) {
      res.end("I don’t have that information in my portfolio data.");
      return;
    }

    if (!contexts.length && (summaryIntent || topic)) {
      res.end("I don’t have that information in my portfolio data.");
      return;
    }

    const prompt = buildPrompt(question, contexts);

    const streamed = await streamGemini(prompt, res);
    if (!streamed) {
      const answer = await callGemini(prompt);
      res.end(answer);
      return;
    }

    res.end();
  } catch (err) {
    res.status(500).json({ error: err.message || "Server error" });
  }
}

File: api/ask.py
import json
import os
import re
import shutil
from pathlib import Path
from typing import Dict, List, Optional, Tuple
from http.server import BaseHTTPRequestHandler

import requests
from langchain_chroma import Chroma
from langchain_openai import OpenAIEmbeddings


ROOT_DIR = Path(__file__).resolve().parents[1]
PERSIST_SRC = ROOT_DIR / "db" / "chroma"
PERSIST_DIR = Path(os.getenv("CHROMA_PERSIST_DIR", "/tmp/chroma"))
COLLECTION = os.getenv("CHROMA_COLLECTION", "kb_docs")
EMBEDDING_MODEL = os.getenv("OPENAI_EMBEDDING_MODEL", "text-embedding-3-small")
GEMINI_MODEL = os.getenv("GEMINI_MODEL", "gemini-1.5-flash")

LAYER_RANK = {"summary": 0, "window": 1, "section": 2, "file": 3, "fallback": 4}

VDB = None


def _ensure_persist_dir() -> Path:
    if PERSIST_DIR == PERSIST_SRC:
        return PERSIST_SRC
    if not PERSIST_DIR.exists():
        if not PERSIST_SRC.exists():
            raise FileNotFoundError(f"Chroma directory not found: {PERSIST_SRC}")
        shutil.copytree(PERSIST_SRC, PERSIST_DIR)
    return PERSIST_DIR


def _get_vectordb() -> Chroma:
    global VDB
    if VDB is not None:
        return VDB
    persist_dir = _ensure_persist_dir()
    embeddings = OpenAIEmbeddings(model=EMBEDDING_MODEL)
    VDB = Chroma(
        collection_name=COLLECTION,
        embedding_function=embeddings,
        persist_directory=str(persist_dir),
    )
    return VDB


def _strip_frontmatter(text: str) -> str:
    text = text.lstrip("\ufeff")
    if not text.startswith("---"):
        return text
    lines = text.splitlines()
    end_idx = None
    for i in range(1, len(lines)):
        if lines[i].strip() == "---":
            end_idx = i
            break
    if end_idx is None:
        return text
    return "\n".join(lines[end_idx + 1 :]).lstrip()


def _normalize_whitespace(text: str) -> str:
    return re.sub(r"\s+", " ", text.strip())


def _infer_topic(query: str) -> Optional[str]:
    q = query.lower()
    if "project" in q:
        return "projects"
    if "experience" in q or "work" in q or "job" in q:
        return "experience"
    if "skill" in q or "stack" in q or "tool" in q:
        return "skills"
    if "education" in q or "degree" in q or "university" in q:
        return "about"
    if "research" in q or "paper" in q:
        return "research"
    return None


def _is_summary_intent(query: str) -> bool:
    q = query.lower()
    triggers = [
        "summarize",
        "summary",
        "overview",
        "who are you",
        "about you",
        "tell me about yourself",
        "background",
    ]
    return any(t in q for t in triggers)


def _build_filter(
    topic: Optional[str],
    layer: Optional[str],
    retrieval_tier: Optional[str],
) -> Dict:
    filters: Dict = {}
    if topic:
        filters["topic"] = topic
    if layer:
        filters["layer"] = layer
    if retrieval_tier:
        filters["retrieval_tier"] = retrieval_tier
    return filters


def _score(distance: float, layer_rank: int, layer_bias: float) -> float:
    return distance + (layer_bias * (layer_rank / 10.0))


def retrieve_context(
    query: str,
    top_k: int = 5,
    fetch_k: int = 15,
    layer_bias: float = 0.15,
) -> Tuple[List[Dict], List[str]]:
    vectordb = _get_vectordb()
    topic = _infer_topic(query)
    layer = "summary" if _is_summary_intent(query) else None
    filters = _build_filter(topic, layer, None) or None

    results = vectordb.similarity_search_with_score(
        query,
        k=fetch_k,
        filter=filters,
    )

    seen_hashes: set[str] = set()
    scored: List[Dict] = []
    for doc, distance in results:
        meta = doc.metadata or {}
        content_hash = meta.get("content_hash")
        if content_hash and content_hash in seen_hashes:
            continue
        if content_hash:
            seen_hashes.add(content_hash)

        layer_rank = int(meta.get("layer_rank", 9))
        adjusted = _score(distance, layer_rank, layer_bias)
        scored.append(
            {
                "score": adjusted,
                "distance": distance,
                "metadata": meta,
                "content": _strip_frontmatter(doc.page_content or ""),
            }
        )

    scored.sort(key=lambda x: x["score"])
    top = scored[:top_k]

    sources: List[str] = []
    for item in top:
        meta = item["metadata"]
        source = meta.get("source_path") or "kb"
        title = meta.get("title") or meta.get("section_title") or "Context"
        sources.append(f"{title} ({source})")

    return top, sources


def build_prompt(query: str, contexts: List[Dict]) -> str:
    parts: List[str] = []
    for item in contexts:
        meta = item["metadata"]
        label = f"{meta.get('topic', 'kb')}/{meta.get('source_path', '')}"
        if meta.get("section_title"):
            label = f"{label} :: {meta.get('section_title')}"
        snippet = _normalize_whitespace(item["content"])
        parts.append(f"[{label}] {snippet}")

    context_block = "\n".join(parts)
    return (
        "Question: " + query + "\n\n"
        "Context:\n" + context_block + "\n\n"
        "Answer clearly and concisely, grounded only in the context."
    )


def call_gemini(prompt: str) -> str:
    api_key = os.getenv("GEMINI_API_KEY")
    if not api_key:
        raise RuntimeError("GEMINI_API_KEY is not set.")

    url = f"https://generativelanguage.googleapis.com/v1beta/models/{GEMINI_MODEL}:generateContent?key={api_key}"
    payload = {
        "system_instruction": {
            "parts": [
                {
                    "text": (
                        "You are Nihal Khan's portfolio assistant. "
                        "Answer only from the provided context. "
                        "If the answer is not in context, say you don't have that information. "
                        "Be recruiter-friendly and concise."
                    )
                }
            ]
        },
        "contents": [{"role": "user", "parts": [{"text": prompt}]}],
        "generationConfig": {"temperature": 0.3, "maxOutputTokens": 512},
    }
    response = requests.post(url, json=payload, timeout=30)
    response.raise_for_status()
    data = response.json()
    candidates = data.get("candidates") or []
    if not candidates:
        return "I don't have enough information to answer that."
    parts = candidates[0].get("content", {}).get("parts", [])
    if not parts:
        return "I don't have enough information to answer that."
    return "".join(part.get("text", "") for part in parts).strip()


class handler(BaseHTTPRequestHandler):
    def do_POST(self) -> None:
        try:
            length = int(self.headers.get("Content-Length", "0"))
            raw = self.rfile.read(length) if length > 0 else b"{}"
            body = json.loads(raw.decode("utf-8"))
            query = (body.get("query") or "").strip()
            if not query:
                self.send_response(400)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(json.dumps({"error": "Missing query"}).encode("utf-8"))
                return

            if not os.getenv("OPENAI_API_KEY"):
                raise RuntimeError("OPENAI_API_KEY is not set.")

            contexts, _sources = retrieve_context(query)
            prompt = build_prompt(query, contexts)
            answer = call_gemini(prompt)

            self.send_response(200)
            self.send_header("Content-Type", "text/plain; charset=utf-8")
            self.end_headers()
            self.wfile.write(answer.encode("utf-8"))
        except Exception as exc:
            self.send_response(500)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"error": str(exc)}).encode("utf-8"))

    def do_OPTIONS(self) -> None:
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()
